### PR TITLE
HOG Descriptor based Object Classification | Find the Break

### PR DESCRIPTION
**Pull request by tessbianchi**
- Adds start for Find The Break Mission. It is not near ready, just a proof of concept for the perception - which is based on KMeans

- Adds changes to Ira code so that stamp is not based on age, but current time.

- Adds object classification system that can use an descriptor and classifier. Currently it is based on HOG Descriptors and a Support Vector Machine. This code structure allows for the to be easily changed, though.

### CODE REVIEW

File: perception/navigator_vision/object_classification/lidar_to_image.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        self.classes = classes
+        self.cam_info = None
+        self.busy = False
+        self.c = 0
+
+        self.debug = Debug(nh)
+
+    @util.cancellableInlineCallbacks
+    def init_(self, right=True):
+        image_sub = "/stereo/right/image_rect_color"
```
Comment (ironmig) [inline perception/navigator_vision/object_classification/lidar_to_image.py]: paramss

File: perception/navigator_vision/object_classification/lidar_to_image.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                                                 lambda odom: setattr(self, 'pose', nt.odometry_to_numpy(odom)[0]))
+        self.cam_info_sub = yield self.nh.subscribe(cam_info, CameraInfo, self._info_cb)
+        self.tf_listener = tf.TransformListener(self.nh)
+        defer.returnValue(self)
+
+    def _info_cb(self, info):
+        self.cam_info = info
+
+    def _odom_cb(self, odom):
+        self.position, self.rot = nt.odometry_to_numpy(odom)[0]
```
Comment (ironmig) [inline perception/navigator_vision/object_classification/lidar_to_image.py]: lambda?

File: perception/navigator_vision/nodes/object_classifier.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        defer.returnValue(self)
+
+
+    def image_cb(self, imgs):
+        for img in imgs:
+            objid = img["id"]
+            img = img["img"]
+            desc = self.descriptor.get_descriptor(img)
+            classification = self.classify(desc)
+            print "id", objid, classification
```
Comment (ironmig) [inline perception/navigator_vision/nodes/object_classifier.py]: I liked the idea of this locking in the classification then setting it in the database here

File: mission_control/navigator_missions/nav_missions/find_the_break.py [hunk 4]
```diff
@@ -1,11 +1,156 @@
...
+    def init_(self):
+        yield self.nh.subscribe("/video_player/croppedFile/image_raw", Image, self.image_cb)
+        defer.returnValue(self)
+
+    def image_cb(self, image):
+        self.image = self.bridge.imgmsg_to_cv2(image, "bgr8")
+
+    @txros.util.cancellableInlineCallbacks
+    def get_direction(self):
+        cap = cv2.VideoCapture("/home/tess/bags/croppedFile.mp4")
```
Comment (ironmig) [inline mission_control/navigator_missions/nav_missions/find_the_break.py]: This seems a little un-rosifed and disconnected from the mission. It should subscribe to an image topic. Maybe you want this perception to be a different class/node?

File: mission_control/navigator_missions/nav_missions/find_the_break.py [hunk 5]
```diff
@@ -1,11 +1,156 @@
...
 @txros.util.cancellableInlineCallbacks
 def main(navigator, **kwargs):
     fprint("STARTING FIND THE BREAK, GETTING A BOUY AS A TEST", msg_color="green")
-    yield navigator.database_query("buoy")
-    yield navigator.nh.sleep(5)
-    fprint("COMPLETED FIND THE BREAK", msg_color="green")
\ No newline at end of file
+    # yield navigator.database_query("buoy")
+    # yield navigator.nh.sleep(5)
+    fprint("COMPLETED FIND THE BREAK", msg_color="green")
```
Comment (ironmig) [inline mission_control/navigator_missions/nav_missions/find_the_break.py]: Shouldn't this go at the end?

File: utils/navigator_tools/navigator_tools/debug.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+import numpy as np
+from sensor_msgs.msg import Image
+from cv_bridge import CvBridge
+import rospy
+import sys
+___author___ = "Tess Bianchi"
+
+
+class Debug(object):
+    """Class that contains methods that assist with debugging with images."""
```
Comment (tessbianchi) [inline utils/navigator_tools/navigator_tools/debug.py]: kk

File: mission_control/navigator_missions/nav_missions/find_the_break.py (hunk 5 above)
Comment (tessbianchi) [inline mission_control/navigator_missions/nav_missions/find_the_break.py]: Yes.

File: mission_control/navigator_missions/nav_missions/find_the_break.py (hunk 4 above)
Comment (tessbianchi) [inline mission_control/navigator_missions/nav_missions/find_the_break.py]: Yes, as it says above, this is just a proof of concept that the perception works, not the actual implementation that will be used on the boat.

File: mission_control/navigator_missions/nav_missions/find_the_break.py [hunk 7]
```diff
@@ -1,11 +1,156 @@
...
+
+    def __init__(self, nh):
+        self.nh = nh
+        self.image = None
+        self.debug = Debug(nh=nh)
+        self.bridge = CvBridge()
+
+    @txros.util.cancellableInlineCallbacks
+    def init_(self):
+        yield self.nh.subscribe("/video_player/croppedFile/image_raw", Image, self.image_cb)
```
Comment (tessbianchi) [inline mission_control/navigator_missions/nav_missions/find_the_break.py]: Yes.

File: perception/navigator_vision/nodes/object_classifier.py (hunk 3 above)
Comment (tessbianchi) [inline perception/navigator_vision/nodes/object_classifier.py]: yes, will be implemented. You can put it in the issue that I made for it for now.

File: .gitignore [hunk 8]
```diff
@@ -42,3 +42,6 @@ Thumbs.db
 .project
 .pydevproject
 .settings
+
+# Kosher Pickles
```
Comment (tessbianchi) [inline .gitignore]: The more ya know

File: perception/navigator_vision/object_classification/lidar_to_image.py (hunk 1 above)
Comment (tessbianchi) [inline perception/navigator_vision/object_classification/lidar_to_image.py]: I would rather use params passed into the function. 

I personally, only thing ros params should be used in between nodes..

File: perception/navigator_vision/object_classification/lidar_to_image.py (hunk 2 above)
Comment (tessbianchi) [inline perception/navigator_vision/object_classification/lidar_to_image.py]: Yes. Gotta think about that line count.

**Comment by tessbianchi** [comment]
This should be good to go. I removed find the break.

**Pull request merged**